### PR TITLE
Fix generic address space OpenCL 2.0 assumption

### PR DESCRIPTION
**Pull request by kbenzie**
Conformance tests should not assume support for OpenCL 2.0, or specific features. This patch removes uses of
`__opencl_c_program_scope_global_variables` for devices which do not support it in the generic address space conformance tests. Additionally, the `clCompileProgram` options have been updated to select between `CL2.0` and `CL3.0` depending on which the device under test supports.

### CODE REVIEW

File: test_conformance/generic_address_space/advanced_tests.cpp [hunk 1]
```diff
@@ -276,31 +304,24 @@ int test_library_function(cl_device_id deviceID, cl_context context, cl_command_
...
-        NL "    failures += helperFunction(lfloatp, lfloat) ? 0 : 1;"
-        NL "    failures += helperFunction(pfloatp, pfloat) ? 0 : 1;"
-        NL
-        NL "    results[tid] = failures == 0;"
-        NL "}"
-        NL;
+    const std::string KERNEL_FUNCTION = NL NL
+        "extern bool helperFunction(float *floatp, float val);" NL NL
+        "#ifdef __opencl_c_program_scope_global_variables" NL
+        "__global float gfloat = 1.0f;" NL "#endif" NL NL
```
Comment (svenvh) [inline test_conformance/generic_address_space/advanced_tests.cpp:310]: The formatting changes don't make this very readable now.  I think you should be able to keep the original layout if you use a C++ raw string literal (and then you won't need `NL` any longer).

**Comment by lakshmih** [comment]
Qualcomm would like more time to review this.

**Review by bashbaug — APPROVED** [review]
At some point it would be nice to factor the code to create a program with source and compile it via `clCompileProgram` into a separate helper function, but I think this is fine for now.

Note that there are not a lot of other places where we call `clCompileProgram` and pass a specific OpenCL C version right now.

File: test_conformance/generic_address_space/advanced_tests.cpp (hunk 1 above)
Comment (kbenzie) [reply test_conformance/generic_address_space/advanced_tests.cpp]: I've made the change to raw string literally for this kernel.

**Review by svenvh — APPROVED** [review]

**Review by lakshmih — APPROVED** [review]
Looks good

Sreelakshmi Haridas Maruthur
Qualcomm

**Pull request merged**